File: ArraysAndStrings/Two Pointers/557ReverseWordsInString3/557ReverseWordsInString3.go
package main

import (
	"fmt"
	"unicode"
)

func reverseWords(s string) string {
	answer := make([]uint8, len(s))

	right := 0
	left := 0

	for right <= len(s)-1 {
		if unicode.IsSpace(rune(s[right])) || right == len(s)-1 {
			answer[right] = s[right]

			fmt.Println(left, right)
			for left < right-1 {
				answer[left] = s[right-left]
				left++
			}
			left++
		}

		right++
	}

	return string(answer)
}

func main() {
	//fmt.Println(reverseWords("Let's take LeetCode contest")) // s'teL ekat edoCteeL tsetnoc
	fmt.Println(reverseWords("Mr Ding")) // rM gniD
}

File: ArraysAndStrings/Two Pointers/557ReverseWordsInString3/557ReverseWordsInString3.py
class Solution:
    def reverseWords(self, s: str) -> str:
        n = len(s)
        result = []
        left = 0
        right = 0

        while left <= n:
            while right < n - 1 and not s[right + 1].isspace():
                right += 1
            new_word = []
            for i in range(right, left-1, -1):
                new_word.append(s[i])
            result.append("".join(new_word))
            right += 2
            left = right

        return " ".join(result)


def test1():
    print(Solution().reverseWords("Let's take LeetCode contest"))  # "s'teL ekat edoCteeL tsetnoc"


def test2():
    print(Solution().reverseWords("Mr Ding"))  # "rM gniD"


test1()
test2()
